File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution(object):
    def shipWithinDays(self, weights, days):
        """
        :type weights: List[int]
        :type days: int
        :rtype: int
        """
        def findDays(n):
            total = 0
            countDays = 1
            
            for i in weights:
                if total + i <= n:
                    total += i
                else:
                    countDays += 1
                    total = i
            return countDays
        
        l, r = -1e9, 0
        
        for i in weights:
            l = max(l,i)
            r += i
        
        if days == 1:
            return r
        
        while l <= r:
            mid = (l+r)//2
            
            if findDays(mid) > days:
                l = mid + 1
            elif findDays(mid) <= days:
                r = mid - 1

        return l
        
        
        

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int l = findMax(weights), r = sumArray(weights);
        int ans = r;
        while (l <= r){
            int mid = l + (r - l) / 2;
            
            if (possible(weights, mid, days)){
                ans = mid;
                r = mid - 1;
            }
            else l = mid + 1;
        }
        return ans;
    }
    
    private boolean possible(int[] weights, int m, int k){
        int days = 0;
        int total = 0;
        for (int w : weights){
            if (total + w > m) {
                days++;
                total = 0;
            }
            total += w;
        }
        days++;
        return days <= k;
    }
    
    private int sumArray(int[] weights){
        int total = 0;
        for (int w: weights){
            total += w;
        }
        return total;
    }
    
    private int findMax(int[] weights){
        int maxi = Integer.MIN_VALUE;
        
        for (int w : weights){
            maxi = Math.max(maxi, w);
        }
        return maxi;
    }
}